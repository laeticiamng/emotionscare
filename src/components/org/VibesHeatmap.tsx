// @ts-nocheck
import React, { useRef, useEffect, useState } from 'react';
import { OrgWeekly, TeamRow, VibeBucket } from '@/hooks/useOrgWeekly';
import { useOrgStore } from '@/store/org.store';
import { AnonymityGuard } from './AnonymityGuard';
import { useTranslation } from 'react-i18next';
import { cn } from '@/lib/utils';

interface VibesHeatmapProps {
  data: OrgWeekly;
  renderMode?: 'svg' | 'canvas';
}

const bucketToColor = (bucket: VibeBucket): string => {
  switch (bucket) {
    case 'low': return 'hsl(var(--vibe-low))';
    case 'medium': return 'hsl(var(--vibe-medium))';
    case 'high': return 'hsl(var(--vibe-high))';
  }
};

const bucketToLabel = (bucket: VibeBucket): string => {
  switch (bucket) {
    case 'low': return 'Bas';
    case 'medium': return 'Stable'; 
    case 'high': return 'Haut';
  }
};

export const VibesHeatmap: React.FC<VibesHeatmapProps> = ({ 
  data, 
  renderMode = 'svg' 
}) => {
  const { t } = useTranslation();
  const { selectedCell, setSelectedCell } = useOrgStore();
  const [focusedCell, setFocusedCell] = useState<{ row: number; col: number } | null>(null);
  
  const eligibleTeams = data.teams.filter(team => team.eligible);
  const dates = eligibleTeams[0]?.days?.map(day => day.date) || [];

  const handleKeyDown = (e: React.KeyboardEvent, row: number, col: number) => {
    let newRow = row;
    let newCol = col;

    switch (e.key) {
      case 'ArrowUp':
        e.preventDefault();
        newRow = Math.max(0, row - 1);
        break;
      case 'ArrowDown':
        e.preventDefault();
        newRow = Math.min(eligibleTeams.length - 1, row + 1);
        break;
      case 'ArrowLeft':
        e.preventDefault();
        newCol = Math.max(0, col - 1);
        break;
      case 'ArrowRight':
        e.preventDefault();
        newCol = Math.min(dates.length - 1, col + 1);
        break;
      case 'Home':
        e.preventDefault();
        newCol = 0;
        break;
      case 'End':
        e.preventDefault();
        newCol = dates.length - 1;
        break;
      default:
        return;
    }

    setFocusedCell({ row: newRow, col: newCol });
    
    // Update selection
    const team = eligibleTeams[newRow];
    const date = dates[newCol];
    if (team && date) {
      setSelectedCell({ teamId: team.team_id, date });
    }
  };

  const formatDate = (dateStr: string): string => {
    const date = new Date(dateStr);
    return date.toLocaleDateString('fr-FR', { weekday: 'short', day: 'numeric' });
  };

  if (eligibleTeams.length === 0) {
    return <AnonymityGuard minN={data.min_n} />;
  }

  return (
    <div 
      className="space-y-4"
      role="table"
      aria-rowcount={eligibleTeams.length + 1}
      aria-colcount={dates.length + 1}
      aria-label="Heatmap des vibes d'équipe"
    >
      {/* Legend */}
      <div className="flex items-center gap-6">
        <span className="text-sm font-medium">Légende :</span>
        <div className="flex items-center gap-4">
          {(['low', 'medium', 'high'] as VibeBucket[]).map((bucket) => (
            <div key={bucket} className="flex items-center gap-2">
              <div 
                className="w-4 h-4 rounded"
                style={{ backgroundColor: bucketToColor(bucket) }}
                aria-hidden="true"
              />
              <span className="text-sm">{bucketToLabel(bucket)}</span>
            </div>
          ))}
        </div>
      </div>

      {/* Heatmap Grid */}
      <div className="overflow-x-auto">
        <div className="min-w-max">
          {/* Header with dates */}
          <div className="grid grid-cols-[200px_repeat(7,80px)] gap-1 mb-2">
            <div></div>
            {dates.map((date) => (
              <div key={date} className="text-xs text-center font-medium p-2">
                {formatDate(date)}
              </div>
            ))}
          </div>

          {/* Team rows */}
          {eligibleTeams.map((team, rowIndex) => (
            <div key={team.team_id} className="grid grid-cols-[200px_repeat(7,80px)] gap-1 mb-1">
              {/* Team name */}
              <div className="flex items-center p-3 bg-muted rounded text-sm font-medium truncate">
                {team.team_name}
                <span className="ml-2 text-xs text-muted-foreground">
                  ({team.size_window})
                </span>
              </div>

              {/* Day cells */}
              {dates.map((date, colIndex) => {
                const dayData = team.days?.find(d => d.date === date);
                const bucket = dayData?.bucket || 'low';
                const isSelected = selectedCell?.teamId === team.team_id && selectedCell?.date === date;
                const isFocused = focusedCell?.row === rowIndex && focusedCell?.col === colIndex;

                return (
                  <button
                    key={date}
                    role="cell"
                    tabIndex={isFocused ? 0 : -1}
                    className={cn(
                      "h-12 rounded border-2 transition-all focus:outline-none focus:ring-2 focus:ring-primary",
                      isSelected && "ring-2 ring-primary",
                      "hover:brightness-110"
                    )}
                    style={{ 
                      backgroundColor: bucketToColor(bucket),
                      borderColor: isSelected ? 'hsl(var(--primary))' : 'transparent'
                    }}
                    aria-label={`${team.team_name}, ${formatDate(date)} : ${bucketToLabel(bucket)}`}
                    onClick={() => setSelectedCell({ teamId: team.team_id, date })}
                    onKeyDown={(e) => handleKeyDown(e, rowIndex, colIndex)}
                    onFocus={() => setFocusedCell({ row: rowIndex, col: colIndex })}
                  />
                );
              })}
            </div>
          ))}
        </div>
      </div>
    </div>
  );
};