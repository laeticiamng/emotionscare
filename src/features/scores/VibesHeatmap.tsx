import { forwardRef, useMemo } from 'react';
import type { ForwardedRef } from 'react';
import { format, startOfISOWeek } from 'date-fns';
import { fr } from 'date-fns/locale';

import type { VibePoint } from '@/services/scores/dataApi';
import {
  buildHeatmapIntensity,
  describeVibe,
  describeVibeIntensity,
  getVibeColor,
} from './verbalizers';

const CELL_SIZE = 22;
const CELL_GAP = 6;
const DAYS = ['Lun', 'Mar', 'Mer', 'Jeu', 'Ven', 'Sam', 'Dim'];

export interface VibesHeatmapProps {
  points: VibePoint[];
  descriptionId: string;
  titleId: string;
}

export const VibesHeatmap = forwardRef<HTMLDivElement, VibesHeatmapProps>(function VibesHeatmap(
  { points, descriptionId, titleId },
  ref: ForwardedRef<HTMLDivElement>,
) {
  const { cells, columnCount } = useMemo(() => buildCells(points), [points]);

  return (
    <figure
      ref={ref}
      data-testid="scores-heatmap-chart"
      role="img"
      aria-labelledby={titleId}
      aria-describedby={descriptionId}
      className="space-y-3"
    >
      <div className="rounded-lg border border-border bg-card p-4">
        <div className="flex gap-4">
          <div aria-hidden="true" className="flex flex-col justify-between py-2 text-xs text-muted-foreground">
            {DAYS.map(day => (
              <span key={day} className="h-[22px] leading-[22px]">
                {day}
              </span>
            ))}
          </div>
          <div
            role="grid"
            aria-label="Vibes quotidiennes"
            className="grid"
            style={{
              gridTemplateColumns: `repeat(${Math.max(columnCount, 1)}, ${CELL_SIZE}px)`,
              gap: `${CELL_GAP}px`,
              gridAutoRows: `${CELL_SIZE}px`,
              gridAutoFlow: 'column',
            }}
          >
            {cells.map(cell => (
              <div
                key={cell.key}
                role="gridcell"
                tabIndex={0}
                className="h-[22px] w-[22px] rounded-md outline-none transition-[box-shadow] focus-visible:ring-2 focus-visible:ring-primary focus-visible:ring-offset-2 focus-visible:ring-offset-background"
                style={{ backgroundColor: cell.fill }}
                aria-label={cell.ariaLabel}
                title={cell.title}
                data-testid={`heatmap-cell-${cell.key}`}
              >
                <span className="sr-only">{cell.title}</span>
              </div>
            ))}
          </div>
        </div>
        <div className="mt-4 flex flex-wrap gap-3" role="list" aria-label="Légende des vibes">
          {['calm', 'focus', 'bright', 'reset'].map(vibeKey => {
            const descriptor = describeVibe(vibeKey);
            if (!descriptor) {
              return null;
            }
            return (
              <div key={vibeKey} className="flex items-center gap-2" role="listitem">
                <span
                  className="inline-flex h-3 w-3 rounded-sm"
                  style={{ backgroundColor: descriptor.palette.medium }}
                  aria-hidden="true"
                />
                <span className="text-xs text-muted-foreground">{capitalize(descriptor.label)}</span>
              </div>
            );
          })}
          <div className="flex items-center gap-2" role="listitem">
            <span className="inline-flex h-3 w-3 rounded-sm bg-muted" aria-hidden="true" />
            <span className="text-xs text-muted-foreground">Aucune vibe détectée</span>
          </div>
          <div className="flex items-center gap-2" role="listitem">
            <div className="flex h-3 overflow-hidden rounded-sm border border-border" aria-hidden="true">
              <span className="h-full w-3 bg-slate-200" />
              <span className="h-full w-3 bg-slate-400" />
              <span className="h-full w-3 bg-slate-600" />
            </div>
            <span className="text-xs text-muted-foreground">Intensité de la vibe</span>
          </div>
        </div>
      </div>
    </figure>
  );
});

interface HeatmapCell {
  key: string;
  fill: string;
  title: string;
  ariaLabel: string;
}

function buildCells(points: VibePoint[]): { cells: HeatmapCell[]; columnCount: number } {
  if (!points.length) {
    return { cells: [], columnCount: 0 };
  }

  const weekColumns = new Map<string, number>();
  const cells: HeatmapCell[] = [];

  points.forEach(point => {
    const date = new Date(point.date);
    if (Number.isNaN(date.getTime())) {
      return;
    }
    const weekStart = startOfISOWeek(date).toISOString();
    if (!weekColumns.has(weekStart)) {
      weekColumns.set(weekStart, weekColumns.size);
    }
    const intensity = buildHeatmapIntensity(point);
    const descriptor = point.vibe ? describeVibe(point.vibe) : undefined;
    const fill = point.vibe ? getVibeColor(point.vibe, intensity ?? 'medium') : '#e2e8f0';
    const label = format(date, "EEEE d MMMM", { locale: fr });
    const vibeLabel = descriptor?.label ? capitalize(descriptor.label) : 'Aucune vibe dominante';
    const intensityLabel = point.vibe ? describeVibeIntensity(intensity) : 'nuance neutre';
    const title = point.vibe ? `${capitalize(label)} — ${vibeLabel} (${intensityLabel})` : `${capitalize(label)} — aucune vibe dominante`;
    const ariaLabel = point.vibe
      ? `${capitalize(label)} : vibe ${vibeLabel}, ${intensityLabel}.`
      : `${capitalize(label)} : aucune vibe dominante.`;
    cells.push({
      key: `${point.date}-${weekColumns.get(weekStart) ?? 0}-${((date.getDay() + 6) % 7)}`,
      fill,
      title,
      ariaLabel,
    });
  });

  return { cells, columnCount: weekColumns.size };
}

function capitalize(value: string) {
  if (!value) {
    return value;
  }
  return value.charAt(0).toUpperCase() + value.slice(1);
}

export default VibesHeatmap;
