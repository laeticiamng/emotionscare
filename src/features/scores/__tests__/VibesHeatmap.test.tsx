// @ts-nocheck
import React from 'react';
import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import VibesHeatmap from '../VibesHeatmap';

describe('VibesHeatmap', () => {
  it('renders a stable accessible grid', () => {
    const { container } = render(
      <VibesHeatmap
        points={[
          { date: '2024-05-06', vibe: 'calm', intensity: 'medium', meta: { count: 2, total: 3 } },
          { date: '2024-05-07', vibe: 'bright', intensity: 'deep', meta: { count: 3, total: 3 } },
          { date: '2024-05-08', vibe: undefined, meta: { count: 0, total: 0 } },
        ]}
        titleId="title"
        descriptionId="desc"
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});

