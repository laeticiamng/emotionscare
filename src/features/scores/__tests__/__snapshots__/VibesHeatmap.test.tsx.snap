// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VibesHeatmap > renders a stable accessible grid 1`] = `
<figure
  aria-describedby="desc"
  aria-labelledby="title"
  class="space-y-3"
  data-testid="scores-heatmap-chart"
  role="img"
>
  <div
    class="rounded-lg border border-border bg-card p-4"
  >
    <div
      class="flex gap-4"
    >
      <div
        aria-hidden="true"
        class="flex flex-col justify-between py-2 text-xs text-muted-foreground"
      >
        <span
          class="h-[22px] leading-[22px]"
        >
          Lun
        </span>
        <span
          class="h-[22px] leading-[22px]"
        >
          Mar
        </span>
        <span
          class="h-[22px] leading-[22px]"
        >
          Mer
        </span>
        <span
          class="h-[22px] leading-[22px]"
        >
          Jeu
        </span>
        <span
          class="h-[22px] leading-[22px]"
        >
          Ven
        </span>
        <span
          class="h-[22px] leading-[22px]"
        >
          Sam
        </span>
        <span
          class="h-[22px] leading-[22px]"
        >
          Dim
        </span>
      </div>
      <div
        aria-label="Vibes quotidiennes"
        class="grid"
        role="grid"
        style="grid-template-columns: repeat(1, 22px); gap: 6px; grid-auto-rows: 22px; grid-auto-flow: column;"
      >
        <div
          aria-label="Lundi 6 mai : vibe Posé, nuance présente."
          class="h-[22px] w-[22px] rounded-md outline-none transition-[box-shadow] focus-visible:ring-2 focus-visible:ring-primary focus-visible:ring-offset-2 focus-visible:ring-offset-background"
          data-testid="heatmap-cell-2024-05-06-0-0"
          role="gridcell"
          style="background-color: rgb(52, 211, 153);"
          tabindex="0"
          title="Lundi 6 mai — Posé (nuance présente)"
        >
          <span
            class="sr-only"
          >
            Lundi 6 mai — Posé (nuance présente)
          </span>
        </div>
        <div
          aria-label="Mardi 7 mai : vibe Lumineux, nuance intense."
          class="h-[22px] w-[22px] rounded-md outline-none transition-[box-shadow] focus-visible:ring-2 focus-visible:ring-primary focus-visible:ring-offset-2 focus-visible:ring-offset-background"
          data-testid="heatmap-cell-2024-05-07-0-1"
          role="gridcell"
          style="background-color: rgb(194, 65, 12);"
          tabindex="0"
          title="Mardi 7 mai — Lumineux (nuance intense)"
        >
          <span
            class="sr-only"
          >
            Mardi 7 mai — Lumineux (nuance intense)
          </span>
        </div>
        <div
          aria-label="Mercredi 8 mai : aucune vibe dominante."
          class="h-[22px] w-[22px] rounded-md outline-none transition-[box-shadow] focus-visible:ring-2 focus-visible:ring-primary focus-visible:ring-offset-2 focus-visible:ring-offset-background"
          data-testid="heatmap-cell-2024-05-08-0-2"
          role="gridcell"
          style="background-color: rgb(226, 232, 240);"
          tabindex="0"
          title="Mercredi 8 mai — aucune vibe dominante"
        >
          <span
            class="sr-only"
          >
            Mercredi 8 mai — aucune vibe dominante
          </span>
        </div>
      </div>
    </div>
    <div
      aria-label="Légende des vibes"
      class="mt-4 flex flex-wrap gap-3"
      role="list"
    >
      <div
        class="flex items-center gap-2"
        role="listitem"
      >
        <span
          aria-hidden="true"
          class="inline-flex h-3 w-3 rounded-sm"
          style="background-color: rgb(52, 211, 153);"
        />
        <span
          class="text-xs text-muted-foreground"
        >
          Posé
        </span>
      </div>
      <div
        class="flex items-center gap-2"
        role="listitem"
      >
        <span
          aria-hidden="true"
          class="inline-flex h-3 w-3 rounded-sm"
          style="background-color: rgb(14, 165, 233);"
        />
        <span
          class="text-xs text-muted-foreground"
        >
          Focalisé
        </span>
      </div>
      <div
        class="flex items-center gap-2"
        role="listitem"
      >
        <span
          aria-hidden="true"
          class="inline-flex h-3 w-3 rounded-sm"
          style="background-color: rgb(249, 115, 22);"
        />
        <span
          class="text-xs text-muted-foreground"
        >
          Lumineux
        </span>
      </div>
      <div
        class="flex items-center gap-2"
        role="listitem"
      >
        <span
          aria-hidden="true"
          class="inline-flex h-3 w-3 rounded-sm"
          style="background-color: rgb(168, 85, 247);"
        />
        <span
          class="text-xs text-muted-foreground"
        >
          Régénérant
        </span>
      </div>
      <div
        class="flex items-center gap-2"
        role="listitem"
      >
        <span
          aria-hidden="true"
          class="inline-flex h-3 w-3 rounded-sm bg-muted"
        />
        <span
          class="text-xs text-muted-foreground"
        >
          Aucune vibe détectée
        </span>
      </div>
      <div
        class="flex items-center gap-2"
        role="listitem"
      >
        <div
          aria-hidden="true"
          class="flex h-3 overflow-hidden rounded-sm border border-border"
        >
          <span
            class="h-full w-3 bg-slate-200"
          />
          <span
            class="h-full w-3 bg-slate-400"
          />
          <span
            class="h-full w-3 bg-slate-600"
          />
        </div>
        <span
          class="text-xs text-muted-foreground"
        >
          Intensité de la vibe
        </span>
      </div>
    </div>
  </div>
</figure>
`;
